### PR TITLE
Add dead features metric

### LINKED ISSUE #88: Log feature density histograms to weights and biases
See [here](https://transformer-circuits.pub/2023/monosemantic-features#:~:text=Specific%20metrics%20from,is%20too%20low.)

To do:
- [ ] Consider the best way to calculate these
- [ ] Log these to wandb (either via wandb histogram or plotly)
- [ ] Also log number of dead features 

There's a few technical details here that could matter but any solution here would be a good start.

### PR DESCRIPTION
**Pull request by lucyfarnik**
Adds histogram and num_dead_features logging to wandb, closes #88

### CODE REVIEW

File: sparse_autoencoder/metrics/histogram_metric.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+from sparse_autoencoder.metrics.train.abstract_train_metric import (
+    AbstractTrainMetric,
+    TrainMetricData,
+)
+
+
+@final
+class HistogramMetric(AbstractTrainMetric):
+    """Histogram metric — log the histogram, as well as the number of dead features."""
```
Comment (alan-cooney) [inline sparse_autoencoder/metrics/histogram_metric.py:15]: Needs a descriptive name - e.g. BatchDeadNeuronsMetric

File: sparse_autoencoder/metrics/histogram_metric.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            log_progress_bar: Whether to log the metric to the progress bar.
+            log_weights_and_biases: Whether to log the metric to Weights and Biases.
+        """
+        super().__init__()
+        self.dead_threshold = dead_threshold
+
+    @final
+    def calculate(self, data: TrainMetricData) -> OrderedDict[str, Any]:
+        """Create a log item for Weights and Biases."""
+        feature_mean_acts = data.learned_activations.mean(dim=0)
```
Comment (alan-cooney) [inline sparse_autoencoder/metrics/histogram_metric.py:35]: I think this should be sum? What I think you're getting at is "how many neurons didn't fire at all in this batch.

File: sparse_autoencoder/metrics/histogram_metric.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        num_dead_features = (feature_mean_acts <= self.dead_threshold).sum().item()
```
Comment (alan-cooney) [inline sparse_autoencoder/metrics/histogram_metric.py:36]: I think this is meaningless with such a small sample size (one batch). However if you just log a table of all neurons (and the indices of these) it may be possible to then do the analysis on wandb? Might be too much data to send.

Alternatively we may need to make this a "resample neurons" metric and log it there rather than for every train batch. Have you experimented with logging this on wandb to see what it looks like?

File: sparse_autoencoder/metrics/l0_norm_metric.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    """L0 (sparsity) norm metric."""
+
+    @final
+    def calculate(self, data: TrainMetricData) -> OrderedDict[str, float]:
+        """Create a log item for Weights and Biases."""
+        # The L0 norm is the number of non-zero elements
+        # (We're averaging over the batch)
+        acts = data.learned_activations
+        value = (torch.sum(acts != 0) / acts.size(0)).item()
+        return OrderedDict(l0_norm=value)
```
Comment (alan-cooney) [inline sparse_autoencoder/metrics/l0_norm_metric.py:24]: Has been added in from the other PR - should be deleted here

File: sparse_autoencoder/metrics/reducer.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        """Get item dunder method."""
+        return self._modules[idx]
+
+    def __iter__(self) -> Iterator[AbstractTrainMetric]:
+        """Iterator dunder method."""
+        return iter(self._modules)
+
+    def __len__(self) -> int:
+        """Length dunder method."""
+        return len(self._modules)
```
Comment (alan-cooney) [inline sparse_autoencoder/metrics/reducer.py:66]: I don't know if we need this. Currently the pipeline just goes through and logs any metrics it's given.

We could abstract it but then maybe don't need all these dunder methods (and as a small point it's a mapping function then, not a reducer).

File: sparse_autoencoder/metrics/tests/test_l0_norm_metric.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    """Test the L0NormMetric."""
+    learned_activations = torch.tensor([[1.0, 0.0, 0.0], [0.0, 0.01, 2.0]])
+    data = TrainMetricData(
+        input_activations=torch.zeros_like(learned_activations),
+        learned_activations=learned_activations,
+        decoded_activations=torch.zeros_like(learned_activations),
+    )
+    log = l0_norm_metric.calculate(data)
+    expected = 2 / 3
+    assert log["l0_norm"] == expected
```
Comment (alan-cooney) [inline sparse_autoencoder/metrics/tests/test_l0_norm_metric.py:25]: Shouldn't be here

**Review by alan-cooney — CHANGES_REQUESTED** [review]
Thanks for adding this!

I've added some comments to help frame it - I think we need to think a bit about where to log & how much of the processing wandb can do. Happy to chat on a call if easier.

**Comment by alan-cooney** [comment]
FYI I'm adding resample metrics here - https://github.com/ai-safety-foundation/sparse_autoencoder/pull/98 - so will pull that part out. I think a metric per batch is still useful but it's different (e.g. maybe we just call it TrainingBatchNeuronActivity ?)
